[dtensor][partial] fixed adding scalar to any Partial

anshul-si opened the pull request:
**Summary:** When we add a scalar to Partial dtensor, we don't redistribute causing us to add the scalar to each part of the partial dtensor as reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This PR addresses this issue so that we always redistribute for Partial when adding a scalar. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

**Test Case**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_partial

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #170632
* #170040
* #170035
* __->__ #170030



cc @wanchaol @tianyu-l @wz337 @XilunWu @d4l3k @pragupta @SherlockNoMad @ppwwyyxx

anshul-si committed:
[dtensor][partial] fixed adding scalar to any Partial

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixed adding scalar to any Partial"

**Summary:** When we add a scalar to Partial dtensor, we don't redistribute causing us to add the scalar to each part of the partial dtensor as reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This PR addresses this issue so that we always redistribute for Partial when adding a scalar. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

**Test Case**
1. pytest test/distributed/tensor/test_pointwise_ops.py -a test_add_scalar_partial




[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixed adding scalar to any Partial"

**Summary:** When we add a scalar to Partial dtensor, we don't redistribute causing us to add the scalar to each part of the partial dtensor as reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This PR addresses this issue so that we always redistribute for Partial when adding a scalar. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

**Test Case**
1. pytest test/distributed/tensor/test_pointwise_ops.py -a test_add_scalar_partial




[ghstack-poisoned]

anshul-si commented on torch/distributed/tensor/_ops/_pointwise_ops.py:
```diff
@@ -531,10 +534,19 @@ def common_pointwise_strategy(
-                partial_supports_linearity = placement.is_partial(
-                    "sum"
-                ) or placement.is_partial("avg")
+                partial_supports_linearity = (
```
My thought process when I read partial_supports_linearity, it's claiming that a Partial (of any type of Partial or any type of reduce_op) will work for given linearity. The safe_avoid_redistribution is another factor that we need to check for is partial_supports_linearity, although I could break up  safe_avoid_redistribution with special_case_add_scalar and not special_case_normpartial.

For this instance, I think we would be considering changing the definition for linearity=1. To be honest, the whole system of linearities only works for regular Partials. If new types of partial placements are added, there's a decent chance that the linearities won't hold up and should be changed. However, if we think that there's low chances of new Partial placements being added, the special cases

ezyang replied:
Collision with https://github.com/pytorch/pytorch/pull/170203 

We need to have this fix in before Wednesday next week for the production run. I don't mind which PR we end up going with

anshul-si committed:
Update on "[dtensor][partial] fixed adding scalar to any Partial"

**Summary:** When we add a scalar to Partial dtensor, we don't redistribute causing us to add the scalar to each part of the partial dtensor as reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This PR addresses this issue so that we always redistribute for Partial when adding a scalar. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

**Test Case**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_partial




[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixed adding scalar to any Partial"

**Summary:** When we add a scalar to Partial dtensor, we don't redistribute causing us to add the scalar to each part of the partial dtensor as reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This PR addresses this issue so that we always redistribute for Partial when adding a scalar. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

**Test Case**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_partial




[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixed adding scalar to any Partial"

**Summary:** When we add a scalar to Partial dtensor, we don't redistribute causing us to add the scalar to each part of the partial dtensor as reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This PR addresses this issue so that we always redistribute for Partial when adding a scalar. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

**Test Case**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_partial




[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixed adding scalar to any Partial"

**Summary:** When we add a scalar to Partial dtensor, we don't redistribute causing us to add the scalar to each part of the partial dtensor as reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This PR addresses this issue so that we always redistribute for Partial when adding a scalar. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

**Test Case**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_partial




[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixed adding scalar to any Partial"

**Summary:** When we add a scalar to Partial dtensor, we don't redistribute causing us to add the scalar to each part of the partial dtensor as reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This PR addresses this issue so that we always redistribute for Partial when adding a scalar. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

**Test Case**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_partial




[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixed adding scalar to any Partial"

**Summary:** When we add a scalar to Partial dtensor, we don't redistribute causing us to add the scalar to each part of the partial dtensor as reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This PR addresses this issue so that we always redistribute for Partial when adding a scalar. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

**Test Case**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixed adding scalar to any Partial"

**Summary:** When we add a scalar to Partial dtensor, we don't redistribute causing us to add the scalar to each part of the partial dtensor as reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This PR addresses this issue so that we always redistribute for Partial when adding a scalar. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

**Test Case**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixed adding scalar to any Partial"

**Summary:** When we add a scalar to Partial dtensor, we don't redistribute causing us to add the scalar to each part of the partial dtensor as reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This PR addresses this issue so that we always redistribute for Partial when adding a scalar. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

**Test Case**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si commented:
@pytorchbot revert

anshul-si commented:
@pytorchbot revert -m "regression in sixlib"

jeanschmidt commented:
@pytorchbot revert -m "regression in sixlib" -c ghfirst

jeanschmidt commented:
@pytorchbot revert -m "regression in sixlib" -c ghfirst

anshul-si committed:
Update on "[dtensor][partial] fixed adding scalar to any Partial"

**Summary:** When we add a scalar to Partial dtensor, we don't redistribute causing us to add the scalar to each part of the partial dtensor as reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This PR addresses this issue so that we always redistribute for Partial when adding a scalar. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

**Test Case**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixed adding scalar to any Partial"

**Summary:** When we add a scalar to Partial dtensor, we don't redistribute causing us to add the scalar to each part of the partial dtensor as reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This PR addresses this issue so that we always redistribute for Partial when adding a scalar. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

**Test Case**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixed adding scalar to any Partial"

**Summary:** When we add a scalar to Partial dtensor, we don't redistribute causing us to add the scalar to each part of the partial dtensor as reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This PR addresses this issue so that we always redistribute for Partial when adding a scalar. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

**Test Case**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]
